x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/parser/ExpressionTests.java

```diff
@@ -282,6 +282,37 @@ public void testCastWithInvalidDataType() {
         assertEquals("line 1:12: Does not recognize type invalid", ex.getMessage());
     }
 
+    public void testCastOperatorWithUnquotedDataType() {
+        Expression expr = parser.createExpression("(10* 2)::long");
+        assertEquals(Cast.class, expr.getClass());
+        Cast cast = (Cast) expr;
+        assertEquals(DataType.INTEGER, cast.from());
+        assertEquals(DataType.LONG, cast.to());
+        assertEquals(DataType.LONG, cast.dataType());
+        assertEquals(Mul.class, cast.field().getClass());
+        Mul mul = (Mul) cast.field();
+        assertEquals("10* 2", mul.sourceText());
+        assertEquals(DataType.INTEGER, mul.dataType());
+    }
+
+    public void testCastOperatorWithQuotedDataType() {
+        Expression expr = parser.createExpression("(10*2)::\"LonG\"");
+        assertEquals(Cast.class, expr.getClass());
+        Cast cast = (Cast) expr;
+        assertEquals(DataType.INTEGER, cast.from());
+        assertEquals(DataType.LONG, cast.to());
+        assertEquals(DataType.LONG, cast.dataType());
+        assertEquals(Mul.class, cast.field().getClass());
+        Mul mul = (Mul) cast.field();
+        assertEquals("10*2", mul.sourceText());
+        assertEquals(DataType.INTEGER, mul.dataType());
+    }
+
+    public void testCastOperatorWithInvalidDataType() {
+        ParsingException ex = expectThrows(ParsingException.class, () -> parser.createExpression("-1::INVALID"));
+        assertEquals("line 1:6: Does not recognize type invalid", ex.getMessage());
```

Comment:
The message needs improving - use "Does not recognize type [{}]", type.sourceText() - to preserve the original declaration inside visitPrimitiveDataType